lib/storage/authgateway.go

```diff
@@ -422,80 +451,128 @@ func (gw *AuthGatewayV2) GetWebPublicAddrs() (addrs []string) {
 }
 
 // SetWebPublicAddrs sets proxy web service public addresses.
-func (gw *AuthGatewayV2) SetWebPublicAddrs(value []string) {
+func (gw *AuthGatewayV1) SetWebPublicAddrs(value []string) {
 	gw.Spec.WebPublicAddr = &value
 }
 
 // CheckAndSetDefaults validates the resource and fills in some defaults.
-func (gw *AuthGatewayV2) CheckAndSetDefaults() error {
+func (gw *AuthGatewayV1) CheckAndSetDefaults() error {
 	if gw.Metadata.Name == "" {
 		gw.Metadata.Name = KindAuthGateway
 	}
 	err := gw.Metadata.CheckAndSetDefaults()
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	if gw.Spec.ConnectionLimits != nil {
+		err := gw.Spec.ConnectionLimits.Check()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	if gw.Spec.Authentication != nil {
+		auth, err := gw.GetAuthPreference()
+		if err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		if auth != nil {
+			err := auth.CheckAndSetDefaults()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+	var addresses []string
+	if gw.Spec.PublicAddr != nil {
```

Comment:
nit: If you had `Spec.PublicAddr []string`, you would not need this check and you'd still be able to tell between unset (`nil`) and empty.

Reply:
Not during marshaling because omitempty skips both empty and nil slices.